### ISSUE TITLE
uniqueness violation clobbers existing database error

### ISSUE THREAD

**Issue by kevinburkeshyp** [balderdashy/sails-postgresql#179]
For constraint or type violations, you get an `err` object that has a lot of useful data about the error - the error code, the table, more detail about the error message, and the constraint that was violated. 

The `formattedErr` created here: https://github.com/balderdashy/sails-postgresql/blob/master/lib/adapter.js#L1232 does not include any of that information, so you have to special case your error handling for uniqueness constraints vs. other types of database errors. In addition it overloads the meaning of the `code` property by assigning `E_UNIQUE` to it (overridden in Waterline to be `E_VALIDATION`.

Another slightly related issue is that Waterline expects there to be a `.model` property set on the error object, but sails-postgresql doesn't set this property. This leads to the confusing error message `"Invalid attributes sent to undefined:"`. I believe if you are using Blueprints this is customer facing.

**Comment by kevinburkeshyp** [comment]
It would also be nice if Waterline used Error objects e.g. `new Error` instead of passing around dictionaries, as Javascript's handling of the latter is unreliable.

**Comment by kevinburkeshyp** [comment]
Also slightly confusing because you only clobber the uniqueness error if the error message matches certain criteria; if you create a partial unique index, for example, you get the normal PG error.

**Comment by particlebanana** [comment]
Yeah `unique` errors could use some love. I hate `WLError` but it's in use too many places to replace the whole thing. Would love some help on getting the pg db errors in line.

**Comment by kevinburkeshyp** [comment]
Turns out if you set `formattedErr.originalError = err` in the `handleQueryError` code, it does what you want!

**Comment by particlebanana** [comment]
@kevinburkeshyp mind submitting a PR?

**Comment by kevinburkeshyp** [comment]
Working on a pull request here but I am not sure how to test it, since (I assume) none of the other libraries implement this interface.

**Comment by kevinburkeshyp** [comment]
Added https://github.com/balderdashy/waterline-adapter-tests/pull/89.

**Comment by sailsbot** [comment]
Thanks for posting, @kevinburkeshyp.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
